Tidy up connection:info command

greg-1-anderson opened the pull request:
The connection:info command was building its own RowOfFields object, whereas an AssociativeList is a much more natural datatype to use in this instance.  Also, failed commands are better off throwing an exception rather than logging an error and returning empty placeholder data.

greg-1-anderson committed:
The connection:info command was building its own RowOfFields object, whereas an AssociativeList is a much more natural datatype to use in this instance.  Also, failed commands are better off throwing an exception rather than logging an error and returning empty placeholder data.

ajbarry reviewed (changes requested):
I think you've accidentally changed the behavior of this command @greg-1-anderson 

The `filter` argument is an output row filter, not a table column filter.

If you spy the feature spec in #1211 you'll see the intent so I hope that doesn't pass once this is merged :-)

SaraMcCutcheon commented on src/Commands/Connection/InfoCommand.php:
```diff
@@ -21,67 +21,48 @@ class InfoCommand extends TerminusCommand
+     *   mysql_port: Mysql Port
+     *   mysql_database: Mysql Database
+     *   mysql_url: Mysql URL
+     *   mysql_command: Mysql Command
```
These "Mysql"s should be "MySQL"

ajbarry replied:
Actually I discussed the labels with @ari-gold and we came to the conclusion that it was better not to relable the parameters in the output as you don't have any programattic way of mapping 'MySQL Command' => `--filter=mysql_command` it would only be implied through reading the help output and assuming a pattern.

greg-1-anderson replied:
@ajbarry Do you mean `--fields=mysql_command`?  Note that if users see `MySQL Command` in the output, they can also use `--fields='MySQL Command'` to filter.

greg-1-anderson replied:
Using the machine names as the column names is, of course, also a perfectly valid option. However, you can also see the machine names quite easily if you use `--format=yaml`.

greg-1-anderson replied:
About the capitalization of the field labels, the current version of Terminus output looks like this:

```
+----------------+------------------------------------------------------------+
| Key            | Value                                                      |
+----------------+------------------------------------------------------------+
| Sftp Username  | dev.ac2022fe-4223-4e80-a290-42a0db87a1d6                   |
| Sftp Host      | appserver.dev.ac2022fe-4223-4e80-a290-42a0db87a1d6.drush.i |
|                | n                                                          |
| Sftp Password  | Use your account password                                  |
| Sftp Url       | sftp://dev.ac2022fe-4223-4e80-a290-42a0db87a1d6@appserver. |
|                | dev.ac2022fe-4223-4e80-a290-42a0db87a1d6.drush.in:2222     |
| Sftp Command   | sftp -o Port=2222 dev.ac2022fe-4223-4e80-a290-42a0db87a1d6 |
|                | @appserver.dev.ac2022fe-4223-4e80-a290-42a0db87a1d6.drush. |
|                | in                                                         |
| Git Username   | codeserver.dev.ac2022fe-4223-4e80-a290-42a0db87a1d6        |
| Git Host       | codeserver.dev.ac2022fe-4223-4e80-a290-42a0db87a1d6.drush. |
|                | in                                                         |
| Git Port       | 2222                                                       |
| Git Url        | ssh://codeserver.dev.ac2022fe-4223-4e80-a290-42a0db87a1d6@ |
|                | codeserver.dev.ac2022fe-4223-4e80-a290-42a0db87a1d6.drush. |
|                | in:2222/~/repository.git                                   |
| Git Command    | git clone ssh://codeserver.dev.ac2022fe-4223-4e80-a290-42a |
|                | 0db87a1d6@codeserver.dev.ac2022fe-4223-4e80-a290-42a0db87a |
|                | 1d6.drush.in:2222/~/repository.git wp-update               |
| Mysql Host     | dbserver.dev.ac2022fe-4223-4e80-a290-42a0db87a1d6.drush.in |
| Mysql Username | pantheon                                                   |
| Mysql Password | 4b33cbdfd64c446788bef86bf9f3af6c                           |
| Mysql Port     | 16191                                                      |
| Mysql Database | pantheon                                                   |
| Mysql Url      | mysql://pantheon:4b33cbdfd64c446788bef86bf9f3af6c@dbserver |
|                | .dev.ac2022fe-4223-4e80-a290-42a0db87a1d6.drush.in:16191/p |
|                | antheon                                                    |
| Mysql Command  | mysql -u pantheon -p4b33cbdfd64c446788bef86bf9f3af6c -h db |
|                | server.dev.ac2022fe-4223-4e80-a290-42a0db87a1d6.drush.in - |
|                | P 16191 pantheon                                           |
+----------------+------------------------------------------------------------+
```

I see that I haven't capitalized things exactly the same as it previously was either. Most places use "URL" rather than "URL", and "MySQL" as you wrote it.

@ari-gold I guess a change in capitalization here is appropriate, i.e. from "Mysql Url" to "MySQL URL"? I know @ajbarry said you wanted to just to use mysql_url here, but that isn't necessarily required simply for the purpose of field selection. Regardless, we should decide and standardize.

SaraMcCutcheon reviewed (changes requested):
There's just that one little thing.

ajbarry commented:
I recommend merging the #1211 and resolving the changes to the feature spec before merging this PR.

greg-1-anderson commented:
@ajbarry You turned the table on its side, and then used `filter` as a row filter. In this PR, I keep the data structure as an associative list with two columns: key and value.

One change of behavior, though, is that `filter` may have also allowed you to select results based on the value field, whereas --fields=foo will only filter based on the key.  If you want to continue to filter on the value, we could enhance the code to do that.

I would suggest that we figure out what sort of filtering we need to support & come up with a good way to do that before implementing a bunch of commands.  Using RowOfFields with `param` and `value` is awkward; we can still filter as needed with AssociativeList.

greg-1-anderson committed:
Correct capitalization, add Redis field labels.

greg-1-anderson commented:
@TeslaDethray I corrected the capitalization of MySQL, as you suggested, and also added the Redis fields I missed.

@ajbarry I did not put back the 'filter' functionality.  Terminus 0.x does not have this, and I am unsure of the utility of filtering on values in this instance.  I think we should do as you said and review the spec before moving forward here.  I would suggest dropping 'filter' unless there's a use-case for it.  The existing --fields option works very similarly to the --field option in Terminus 0.x.

Also, speaking of --field, I thought I might add --field back in and make `--field=foo` a synonym for `--fields=foo --format=string`.  This would re-introduce the behavior of --field in a more-or-less identical-to-0.x form.

Here is what I mean:

```
$ php ./bin/terminus.php connection:info wp-update.dev --fields=git_port
 ---------- ------ 
  Git Port   2222  
 ---------- ------ 
$ php ./bin/terminus.php connection:info wp-update.dev --fields=git_port --format=string
2222
```

So, `terminus connection:info --field=git_port` would be like the ~~session~~ **second** example.

Finally, there's one more thing that needs to be considered.  When I added the Redis field labels, this causes the default table to always include the Redis fields. This is because default-fields for output formatters is "all of the fields", unless a different list is provided via `@default-fields`. However, if we added a `@default-fields` that was missing the Redis fields, then they wouldn't show up unless the user specifically added them via --fields, which would be tedious.

Perhaps default fields should be filtered such that any item missing in the output data is removed from the resulting display.  This would in general be better, unless there are use-cases where you'd want the default field to be there, but empty.  (If the user specifically requests a field, then it will be there but empty if it has no value.)

ajbarry commented:
@greg-1-anderson I am +1 on keeping the behavior closer to 0.x at this point and not concerned with the mechanics of how that is implemented.

... and to immediately go against that recommendation :-) ...

@ari-gold was asking if we could/should have the default fields be: `sftp_command, git_command, mysql_command, redis_command` arguing that it would be nice and succinct for our command-line just provide you with the other command-lines to run. I would support `@default_fields` for that case. Less concerned if the field is empty in the redis case... `null` or `""` is still a valid value.

greg-1-anderson commented:
Making the default fields *_command makes sense to me.

greg-1-anderson committed:
Update to latest output formatters library. Add default fields ('*_command'). Update tests.

greg-1-anderson committed:
Add a couple more field-format tests.

Commit:
Merge branch 'master' into connection-info-asociative-list

SaraMcCutcheon commented:
It's still beautiful, @greg-1-anderson. Re+1
